src/mpid/ch4/shm/xpmem/xpmem_control.c

```diff
@@ -120,11 +105,125 @@ int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_send_fin_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq = (MPIR_Request *) ctrl_hdr->xpmem_slmt_send_fin.req_ptr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_SEND_FIN_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_SEND_FIN_CB);
+
+    XPMEM_TRACE("send_lmt_send_fin_cb: complete sreq %p\n", sreq);
+    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(sreq, datatype));
+    MPID_Request_complete(sreq);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_SEND_FIN_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_recv_fin_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *rreq = (MPIR_Request *) ctrl_hdr->xpmem_slmt_recv_fin.req_ptr;
+    MPIDI_XPMEM_coop_dmsg_t *dmessage;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_RECV_FIN_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_RECV_FIN_CB);
+
+    XPMEM_TRACE("send_lmt_recv_fin_cb: complete rreq %p\n", rreq);
+
+    if (MPIDI_XPMEM_REQUEST(rreq, counter)) {
+        MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+        OPA_store_int(MPIDI_XPMEM_REQUEST(rreq, counter), 0);
+    }
+
+    MPID_Request_complete(rreq);
+
+    if (MPIDI_XPMEM_global.coop_dmsg_queue) {
+        MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t *slmt_rts_hdr;
+        dmessage = MPIDI_XPMEM_global.coop_dmsg_queue;
+        slmt_rts_hdr = &dmessage->ctrl_hdr.xpmem_slmt_rts;
+        if (!OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                          [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                           slmt_rts_hdr->src_lrank])) {
+            DL_DELETE(MPIDI_XPMEM_global.coop_dmsg_queue, dmessage);
+            mpi_errno = MPIDI_XPMEM_ctrl_do_send_lmt_rts_cb(&dmessage->ctrl_hdr);
+            MPIR_ERR_CHECK(mpi_errno);
+
+            MPL_free(dmessage);
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_RECV_FIN_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_rts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_coop_dmsg_t *dmessage;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t *slmt_rts_hdr =
+        &((MPIDI_SHM_ctrl_hdr_t *) ctrl_hdr)->xpmem_slmt_rts;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_CB);
+
+    if (OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                     [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                      slmt_rts_hdr->src_lrank])) {
+        /* Previous copy is not done, need to delay current copy to next round. */
```

Comment:
I don't understand why second `mpi_isend(1M)` would be put in `coop_dmsg_queue`. Since your first `mpi_isend(1M)` has been put in unexpected queue, the counter should be 0 when second `mpi_isend(1M)` message arrives. 

When receiver uses up the counter and dynamically allocate, the sender needs to perform `xpmem_attach` that is huge overhead in that case. We have considered it before I implemented `dmsg_queue`. If it's hard to predict how many message that will arrive in the future, and attaching overhead is also not predictable.

Reply:
Let me adjust the example a bit.
```
mpi_isend(1M)-> enqueue unexp_rreq
                              mpi_irecv(1M) // matched with 1st isend
                              -> performing coop copy
mpi_isend(1M)-> enqueue coop_dmsg_queue
mpi_isend(1) -> enqueue unexp_rreq
                              mpi_irecv(1M) // matched with 3rd isend
                              ...
```
To fix this program, we need to enqueue the 2nd isend into `unexp_rreq` before enqueuing into `coop_dmsg_queue`. Consider this way, we need to check whether coop copy is ready only when the message has been actually matched with a receive request.

Reply:
The reason I suggested dynamic counter pool is because of two reasons:
1. It is hard to maintain the correctness of message matching (see the above example).
2. It disables parallel data transfer when two threads on a process are communicating with a remote process simultaneously (see the comment in `xpmem_init`).

The first issue can be fixed, but the second issue seems to be a fundamental drawback of the fixed-counter design. Suppose we go with the dynamic counter pool design, the first `X` counters on each process are still statically allocated/attached at init time, thus I do not think it gives more overhead than the current design. What it can additionally provide is that, (1) the messages don't have to be queued when pre-allocated counters are used up;(2) the counters on each process can be shared by all local peers, thus more efficient for the cases where communication is not balanced.